proto/docs/p4runtime.md

```diff
@@ -515,3 +515,164 @@ messages refer to singleton ports by default.
+above annotations defined.
+
+The sub-sections below provide comprehensive examples for various translation
+scenarios.
+
+### Translation of Header Fields
+Port and class of service type fields can be part of header types. For example,
+a packet-out header from the controller to the switch may define the egress
+port of the packet as shown below.
+```
+@controller_header("packet_out")
+header PacketOut_t {
+  @p4runtime_translation("port")
+  PortIdInHeader_t egress_port;
+}
+```
+The top annotation `@controller_header` is also a standard P4Runtime annotation
+that identifies a header type for a controller packet-out. When the target
+receives a packet-out from the controller over the P4Runtime stream channel, it
+will normally expect packet-out metadata (`egress_port`) value of bitwidth as
+defined by the PSA target's `PortIdInHeader_t` type and in the space of the
+target's port numbers. However, the presence of the `@p4runtime_translation`
+annotation will signal to the p4runtime service that it should expect a
+controller defined type and value of `egress_port` metadata. The p4runtime
+service should use the target config to translate this value into PSA
+device-specific value of type `PortIdInHeader_t`.
+
+A similar reverse translation is required in the P4Runtime service for packets
+punted from the target to the controller as shown by the packet-in header
+example below.
+```
+@controller_header("packet_in")
+header PacketIn_t {
+  @p4runtime_translation("port")
+  PortIdInHeader_t ingress_port;
+  @p4runtime_translation("cos")
+  ClassOfServiceInHeader_t cos;
+}
+```
+A packet punted from the target's PSA device will be intercepted by the
+P4runtime service before being sent to the controller. Normally, the P4Runtime
+service will extract the values of `ingress_port` and `cos` from the header and
+insert them in the packet-in metadata fields before sending the packet over the
+stream channel to the controller. However, the presence of the
+`@p4runtime_translation` annotations will be a signal to the P4Runtime service
+to replace these values with equivalent controller-defined numbers from the
+target config.
+
+### Translation of Match Fields
+Port and class of service type match fields can be part of a P4 table's match
+key as shown in the example below:
+```
+table t {
+  key = {
+    istd.ingress_port: exact @p4runtime_translation("port");
+    istd.class_of_service: ternary @p4runtime_translation("cos");
+  }
+  actions = {
+    drop;
+  }
+}
+```
+Table `t` has an `exact` match on ingress port and `ternary` match on class of
+service metadata. A P4Runtime write request from the controller will have
+the values of the match fields set to controller-specific values. The
+`@p4runtime_translation` annotation on the match fields is a signal to the
+P4Runtime service to intercept these write requests and use the configuration
+data to translate the values to respective device-specific values. Similarly,
```

Comment:
I would like to see an expansion of the implications of this translation as they pertain to the "Byte strings" section in the early part of the document.  The PSA spec implies that a type of unspecified width should be treated as 64-bits wide in the P4Runtime API (section 4.1).  Thus, it seems like the P4Runtime service should expect the controller to provide an 8-byte/64-bit string when specifying the match value for istd.ingress_port in table t.  It also seems that the translation can introduce additional errors, depending on how it is implemented and the configuration data provided to the translation logic.  Does the service need to differentiate between pre-translation and post-translation errors?  Is there a difference between a controller-provided string that is a known bad input to the translation logic and another string that fails to translate once it passes as a valid input?

Reply:
Great points Tom! Added clarification.